### PR TITLE
Sync workalendar 14.1.0

### LINKED ISSUE #17: Sync with Workalendar 14
Calendra last synced with Workalendar at version 8, so is due for a refresh.

### PR DESCRIPTION
**Pull request by ShaheedHaque**
Resolves #17. All tests clean for v14.1.0, please consider merging and releasing.

### CODE REVIEW

**Comment by ShaheedHaque** [comment]
@jaraco Hi, if you don't have time to look at this, would you perhaps be prepared to make me a contributor and able to make releases?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
+name: CI
```
Comment (jaraco) [inline .github/workflows/ci.yml:1]: We'll want to delete this file; the 'main' workflow from calendra should be sufficient.

File: CHANGES.rst [hunk 2]
```diff
@@ -1,15 +1,219 @@
...
+
+- Fix several miscalculations in Georgia (USA) calendar (#451).
+
+v8.0.1 (2020-01-24)
 
-Fix version inference when installed from sdist.
+- Fix Family Day for British Columbia (Canada) which was switched from 2nd to 3rd Monday of February in 2019 - thx @jbroudou for the bug report (#454).
 
-v6.1.0
+v8.0.0
```
Comment (jaraco) [inline CHANGES.rst:216]: Calendra follows its own versioning scheme, so since 6.1.0 was the last release, probably 7.0.0 will be the next release, and it will incorporate all of the changes from workalendra 8.0.1-14.1.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ShaheedHaque) [reply .github/workflows/ci.yml:1]: Cool. Want to do the honours, or shall I?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
+# Workalendar
```
Comment (jaraco) [inline README.md:1]: The renaming (and divergent syntax) of the README is going to be a difficult divergence to manage.

This project will need to decide if it wants to adopt the .md format or continue to support the .rst format. I'm much more comfortable with the .rst format, because that's what my other projects use and because it's the default for Python projects. I prefer the md format for it's more universal acceptance. At this point, I don't feel strongly enough about either format to advocate for either, so I'll defer to your preference. If you have no preference, let's go with md, as I think that's the format that will be preferable going forward.

The important thing is going to be maintaining the ancestry and reconciling the changes across the file rename. Here's what I suggest:

- First merge with workalendar at the point right before the rename commit.
- In calendra, rename the file using `git rename`.
- Merge with workalendar after the rename commit and in this merge commit or in a subsequent commit,
- Update any reStructuredText syntax with Markdown syntax.

File: CHANGES.rst (hunk 2 above)
Comment (ShaheedHaque) [reply CHANGES.rst:216]: I'm easy either way, but I can see some small advantage in keeping comparable numbers. Again, feel free to revise them if you like, or I'm happy to comply too.

File: README.md (hunk 3 above)
Comment (ShaheedHaque) [reply README.md:1]: I'm cool with staying on .md, and agree its a right pain to keep in sync. But I'm not clear how to do what you suggest with just this one file (i.e. rewinding the merge would be a LOT of work).

File: calendra/core.py [hunk 4]
```diff
@@ -506,12 +526,231 @@ def get_working_days_delta(self, start, end, include_start=False):
...
+        if self.include_new_years_day and self.shift_new_years_day:
+            if new_year.weekday() in self.get_weekend_days():
+                days.append((
+                    self.find_following_working_day(new_year),
+                    "New Year shift"))
+        return days
+
+
+class SeriesShiftMixin:
+    """
```
Comment (ShaheedHaque) [inline calendra/core.py:689]: @jaraco I wanted to draw your attention to this new Mixin. Hopefully, I explained it's purpose but if I was not clear enough, please ask.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ShaheedHaque) [reply .github/workflows/ci.yml]: Done

**Comment by jaraco** [comment]
> Resolves #15. All tests clean for v14.1.0, please consider merging and releasing.

I suspect this issue doesn't affect #15. I see there's already a test for 2020-12-28 as an observed holiday, so I'll follow up in that ticket.

Still, I think we want to do this work to fix #17.

> @jaraco Hi, if you don't have time to look at this, would you perhaps be prepared to make me a contributor and able to make releases?

Thanks for the ping. I don't subscribe to my own repos (too much noise, so I need to batch my work), but thanks for the mention to remind me (otherwise could have been months ;).

File: CHANGES.rst (hunk 2 above)
Comment (ShaheedHaque) [reply CHANGES.rst:216]: I looked at the file, and am concerned that having two sets of version numbers will be even more confusing to read. At least the current approach avoids that...

Comment (jaraco) [reply CHANGES.rst:216]: The problem I've had in the past was that if calendra wanted to make a change independent of workalendar, it would cause an increment in version numbers. To honor semver, that increment might be a major (1.x) bump. So then calendra is constrained to only commit changes at the same scope as (and coordinated with) workalendar changes or its version numbers must get ahead of workalendar. In short, there's no way to honor semver and keep the numbers in sync. That's why I decided to honor semver and use an otherwise independent versioning scheme.

I use the "Incorporate changes from workalendar vN.N.N" in the changelog to provide a clear and direct association of changes inherited from the upstream. Let's keep that scheme for now.

File: README.md (hunk 3 above)
Comment (ShaheedHaque) [reply README.md:1]: I'll take a stab at merging the Calendra-flavoured content by hand into the .md.

Comment (jaraco) [reply README.md:1]: > But I'm not clear how to do what you suggest with just this one file

Let me take a stab at it. I don't want Git to think these are two different files, else changes in the skeleton to README.rst will show as conflicted with a deleted file instead of conflicted with README.md.

**Comment by jaraco** [comment]
> would you perhaps be prepared to make me a contributor and able to make releases

Most definitely. Your help on this project has been immeasurably helpful. I even tried making this merge recently and gave up before completing it. Let's work through this PR together and after, I'll grant access to help maintain the project and guide on cutting a release.

File: CHANGES.rst (hunk 2 above)
Comment (ShaheedHaque) [reply CHANGES.rst:216]: ok, i'll work this in.

File: README.md (hunk 3 above)
Comment (jaraco) [reply README.md:1]: I notice that `checkdocs` doesn't work with md-formatted readme.

Comment (jaraco) [reply README.md:1]: I'm working on a new release of pytest-checkdocs that will at least not fail on markdown content.

Comment (jaraco) [reply README.md]: Fixed in 8f54646168a6e8e665557b93a20661d0ec7eb623.

File: calendra/core.py (hunk 4 above)
Comment (jaraco) [reply calendra/core.py:689]: Thanks for pointing this out. Is this change necessary to support the newer workalendar? It would be nice to consider this new feature (and divergence from workalendar) independently from the sync with workalendar. If it's not strictly necessary to get a synced version to work, then I'd recommend considering this separately from the rest of the sync. If it is necessary, would it be possible to contribute this functionality to calendra master prior to syncing with workalendar?

The current calendra already deals with some small series (Thanksgiving in US, Christmas+Boxing day in UK). I see Thanksgiving doesn't bother with shifts because it always falls on a consistent day of the week. But [here is how Boxing Day indicates that it falls after Christmas](https://github.com/jaraco/calendra/blob/fa03ff090d39e85ed2099571e14b1e8f3ec17944/calendra/core.py#L653). In this way, Boxing day knows to honor both Christmas and its shift relative to any given calendar.

Ideally, a "Series" would be handled similarly - where each day is a Holiday, and each holiday is aware of its dependencies. Even better might be to reflect the reality of the situation and let a single Holiday be observed on multiple days (and to allow that holiday to set its rules for when it's observed, relative to the configuration of the calendar).

In other words, this hook point feels like not the best place to model the behavior. I'd hope that there could be a cleaner way than keeping track of series by label and copy/pasting the observance shift from each.

I'll have to think about this some more, but please let me know what you think about my questions above.

Comment (jaraco) [reply calendra/core.py:689]: Although I've expressed some reluctance here, I suspect this approach may be suitable at least for an initial implementation. Please answer the questions if you can, but don't feel discouraged. Let's focus on getting the tests passing and iterate from there.

**Comment by jaraco** [comment]
It looks like the tests are failing on Windows, all the ones about ical export.

**Review by jaraco — APPROVED** [review]

File: calendra/core.py (hunk 4 above)
Comment (ShaheedHaque) [reply calendra/core.py:689]: Good questions, let me start from where I found the issue...

When trying to sync up to the upstream v11.0.0 release, I was getting test failures for China and some Islamic countries (Turkey, Malaysia adn Qatar). While poking around these Islamic failures, I realised that the two Eid holiday "series" were not being shifted correctly: only the day in the series that actually fell on a Sunday was being shifted, and that shift then caused it to be a duplicate of the next day in the Series.

This then fell foul of the Workalendar support for ical export (introduced circa Aug 2020). This includes tests which assert, amongst other things, on the uniqueness number of `uid_lines`...which the duplicates now broke.

I realised at this point that I could not see any logic in either Workalendar or Calendra to deal with this - except that I had already seen the custom logic that handled Christmas day and Boxing day (recall issue #15 is why I was here in the first place).

Note: since we have your observance day logic as distinct from what upstream does  (which is why I choose to use Calendra over Workalendar), there are two different codebases with the same problem! I immediately realised that I had little prospect of a decent/viable fix for upstream, but your logic provides easy detection AND a basis for a somewhat tacky solution. The current implementation is a direct consequence of this train of thought.

(Once I implemented it for Malaysia, I realised the same fix was needed for the other Islamic countries, and shortly later for the various series holidays in China).

The shortcomings of the solution are:

1. No obvious backport upstream.
2. As you point out, the lack of a code structure corresponding to the problem in hand.  Conceptually at least, this is broken since it cannot handle the obvious possibility that the series itself may have a knock on effect on a later holiday.
3. No immediate backport to the Christmas/Boxing day series. This is a direct consequence of the conceptual problem from 2.
4. I especially don't like the need to tailor the copy-pasting of the observance for each year in case the shift point moves, and all the duplication of Holiday. (Part of that at least could be optimised but...).

I felt obliged to ignore 1, and the solution for 3 would have been intrusive. As for 2, I considered subclass of Holiday to wrap the series, but could not see a comfortable fit for the observed_date() API:

- A covariant return like `Union[date, List[date]]` seemed horrid
- A new observed_dates() API seemed worse.

So that's why I ended up with what you see.

Finally, with hindsight, it would definitely have been better as a separate commit ahead of the sync. But I didn't think about that at the time, and not having the fix meant that v11.0.0's tests would not pass. Moving it now would be tedious, but may be the right thing to do (not least because it would make filing a bug upstream easier). Your call...

Comment (jaraco) [reply calendra/core.py:689]: Thanks for the response Shaheed. Your clarity and diligence and eloquence in helping me understand the situation is exceptional.

**Comment by ShaheedHaque** [comment]
I just had a thought about the series stuff. The implicit assumption is
that if one of the days is shifted then the days after it should be shifted
too (otherwise the original shift was meaningless).

However, in each case I recall, the original shift is based on a country
level observance rule (usually involving skipping Sunday I think). If that
original shift was incorrect, then clearly applying the series shifts would
also be incorrect.

Is it possible that any of the original country shifts was not right for
the holidays involved?
